devices/src/pvmemcontrol.rs

```diff
@@ -0,0 +1,833 @@
+    }
+
+    fn ready(PercpuInitState { port_buf_map, .. }: PercpuInitState) -> Self {
+        PvmemcontrolDevice::new(
+            PvmemcontrolTransport::ack(),
+            PvmemcontrolState::Ready(port_buf_map),
+        )
+    }
+
+    fn run_command(
+        &mut self,
+        guest_memory: &GuestMemoryAtomic<GuestMemoryMmap<AtomicBitmap>>,
+        command: PvmemcontrolTransportCommand,
+    ) {
+        let state = std::mem::replace(&mut self.state, PvmemcontrolState::Broken);
+
+        *self = match command {
+            PvmemcontrolTransportCommand::Reset => Self::reset(),
+            PvmemcontrolTransportCommand::Register => {
+                if let PvmemcontrolState::PercpuInit(state) = state {
+                    Self::register_percpu_buf(guest_memory, state, unsafe {
+                        self.transport.as_register()
+                    })
+                } else {
+                    warn!("received register without reset");
+                    Self::error()
+                }
+            }
+            PvmemcontrolTransportCommand::Ready => {
+                if let PvmemcontrolState::PercpuInit(state) = state {
+                    Self::ready(state)
+                } else {
+                    warn!("received ready without reset");
+                    Self::error()
+                }
+            }
+            PvmemcontrolTransportCommand::Disconnect => Self::error(),
+            PvmemcontrolTransportCommand::Ack => {
+                warn!("received ack as command");
+                Self::error()
+            }
+            PvmemcontrolTransportCommand::Error => {
+                warn!("received error as command");
+                Self::error()
+            }
+        }
+    }
+
+    /// read from the transport
+    fn read_transport(&self, offset: u64, data: &mut [u8]) {
+        self.transport
+            .as_slice()
+            .iter()
+            .skip(offset as usize)
+            .zip(data.iter_mut())
+            .for_each(|(src, dest)| *dest = *src)
+    }
+
+    /// can only write to transport payload
+    /// command is a special register that needs separate dispatching
+    fn write_transport(&mut self, offset: u64, data: &[u8]) {
+        self.transport
+            .payload
+            .as_mut_slice()
+            .iter_mut()
+            .skip(offset as usize)
+            .zip(data.iter())
+            .for_each(|(dest, src)| *dest = *src)
+    }
+
+    fn find_connection(&self, conn: GuestConnection) -> Option<GuestAddress> {
+        match &self.state {
+            PvmemcontrolState::Ready(map) => map.get(&conn).copied(),
+            _ => None,
+        }
+    }
+}
+
+pub struct PvmemcontrolBusDevice {
+    mem: GuestMemoryAtomic<GuestMemoryMmap<AtomicBitmap>>,
+    dev: RwLock<PvmemcontrolDevice>,
+}
+
+pub struct PvmemcontrolPciDevice {
+    id: String,
+    configuration: PciConfiguration,
+    bar_regions: Vec<PciBarConfiguration>,
+}
```

Comment:
Would you like to explain why do you need use two structs to represent the device? In my opinion, one 
 struct maybe PvmemcontrolDevice seems like enough.

Reply:
Right. My observation was that both BusDevice and PciDevice handle device writes/reads, but actually only the BusDevice impl received the writes/reads. I want the device to handle requests on multiple cpus at the same time. So I made the BusDeviceSync trait similar to crosvm, which is just the BusDevice trait without the exclusive ref requirement on the read and write trait methods so the impl can handle its own locking, so multiple read locks can be taken at the same time.

I left the PciDevice trait in place, so I need two structs because the PciDevice gets wrapped in an Arc<Mutex<>> when I want a RwLock. On second thought, maybe I should instead refactor the Pci/BusDevice traits such that PciDevice also handles its own locking? That would be more consistent, but also inflate the PR to a tree-wide change.